src/main/java/com/google/devtools/build/lib/remote/AbstractActionInputPrefetcher.java

```diff
@@ -224,14 +224,24 @@ protected AbstractActionInputPrefetcher(
   }
 
   private static boolean shouldDownloadFile(Path path, FileArtifactValue metadata) {
-    if (!path.exists()) {
+    var stat = path.statNullable();
+    if (stat == null) {
       return true;
     }
 
-    // In the most cases, skyframe should be able to detect source files modifications and delete
-    // staled outputs before action execution. However, there are some cases where outputs are not
-    // tracked by skyframe. We compare the digest here to make sure we don't use staled files.
+    // In most cases, skyframe should be able to detect source files modifications and delete
+    // stale outputs before action execution. However, there are some cases where outputs are not
```

Comment:
@coeuvre I'm curious which files you had in mind when you introduced this in https://github.com/bazelbuild/bazel/commit/ebd6e58ac186d5137a4ba4450ae48a28b1e111f7. Is this perhaps about spawn outputs that aren't action outputs (e.g. `test.xml`)? If so, could we optimize this further by skipping the checks for non-test actions?

Reply:
In that case, what do you think of adding a check for whether the artifact is an output of the action (which we have available) and skipping I/O in that case? Even the stat in the beginning shouldn't be necessary in that case.